client/snapctl.go

```diff
@@ -62,16 +62,23 @@ type snapctlOutput struct {
 	Stderr string `json:"stderr"`
 }
 
+// protect against too much data via stdin
+var stdinReadLimit = int64(4 * 1000 * 1000)
+
 // RunSnapctl requests a snapctl run for the given options.
 func (client *Client) RunSnapctl(options *SnapCtlOptions, stdin io.Reader) (stdout, stderr []byte, err error) {
 	// TODO: instead of reading all of stdin here we need to forward it to
 	//       the daemon eventually
 	var stdinData []byte
 	if stdin != nil {
-		stdinData, err = ioutil.ReadAll(stdin)
+		limitedStdin := &io.LimitedReader{R: stdin, N: stdinReadLimit}
```

Comment:
Unfortunately LimitedReader does not make it easy to detect when it hit the limit if one is not invoking Read() manually and inspecting N afterwards. So this should be stdinReadLimit + 1 to have a 4MB limit with the current logic AFAIU.

Reply:
Hmm I'm not sure this is true unless I'm missing something... ioutil.ReadAll is a wrapper for io.ReadAll which wraps Read; LimitedReader's Read() is guaranteed to stop reading with EOF upon reaching N; N can be inspected afterwards.

https://golang.org/src/io/io.go?s=15540:15633#L451

Reply:
When the input is exactly N bytes, N will be 0 too, and AFAICT it never goes < 0, which looks identical to a case when there was more bytes in the input and LimitedReader returned EOF early. See: https://play.golang.org/p/MPlRZmrx11s So you cannot really distinguish a case from a valid input of exactly N bytes, and too long one that was trimmed.

Reply:
Ahhh, got you, thanks for the example, that's a bummer /o\